Sources/BSON/Document.swift

```diff
@@ -9,55 +36,115 @@ public struct Document {
 
     private var _buffer: ByteBuffer
 
+    internal var size: Int {
+        guard let size = self.buffer.getInteger(at: 0, endianness: .little, as: Int32.self) else {
+            return 5
+        }
+        return Int(size)
+    }
+
     /// An unordered set containing the keys in this document.
     private var keySet: Set<String>
 
     internal init(_ elements: [BSON]) { fatalError("Unimplemented") }
 
-    internal init(keyValuePairs: [(String, BSON)]) { fatalError("Unimplemented") }
+    internal init(keyValuePairs: [(String, BSON)]) {
+        self.keySet = Set(keyValuePairs.map { $0.0 })
+        guard self.keySet.count == keyValuePairs.count else {
+            fatalError("Dictionary \(keyValuePairs) contains duplicate keys")
+        }
+
+        self._buffer = BSON_ALLOCATOR.buffer(capacity: 100)
+
+        guard !self.keySet.isEmpty else {
+            // 5 byte count as Int32 + null byte
+            self._buffer.writeBytes([5, 0, 0, 0] + [0])
```

Comment:
I think it would be more explicit to do this the way you did in the empty initializer below where you write the int32 5 and then write the null byte. (maybe you can even delegate to that initializer from here?)